Make compatible with ClojureScript

danieroux opened the pull request:
This makes semantic-ui-wrapper usable by ClojureScript, while keeping compatibility with shadow-cljs.

This support requires ClojureScript v1.10.844, released April 2021: https://clojurescript.org/news/2021-04-06-release#_library_property_namespaces 

Some minor modifications to the documentation being generated are also present. 

I kept the three commits separate, to split the generation from the changes.

awkay commented on project.clj:
```diff
@@ -1,35 +1,71 @@
+                 ; [org.clojure/clojurescript "1.10.439" :scope "provided"]
+                 [fulcrologic/fulcro "2.6.18" :scope "provided"]
+                 [cljsjs/semantic-ui-react "0.84.0-0"]
+                 [nubank/workspaces "1.0.2"]
```
workspaces should be a dev dependency, and shadow-cljs should not be a dependency at all.  This is a library.  Do not cause transitive dependency requirements for the consumers of it.

awkay commented on shadow-cljs.edn:
```diff
@@ -0,0 +1,37 @@
+{:lein   {:profile "cljs,dev"}
+ :nrepl  {:port 9000}
+ :builds {
+          ; :main       {:target     :browser
```
Why are you checking in commented source?

awkay commented on project.clj:
```diff
@@ -1,35 +1,71 @@
-                                                   :output-to            "resources/public/js/cards.js"
-                                                   :preloads             [devtools.preload]
-                                                   :source-map-timestamp true}}]}
+             ; :dev {:source-paths ["src/main" "src/cards"]
```
Don't check in commented out source.  Either it belongs or it doesn't. Git can track the old versions

awkay commented on src/dev/user.clj:
```diff
@@ -0,0 +1,186 @@
+
+(set-refresh-dirs "src/main" "src/test" "src/cards" "src/workspaces")
+
+;FIGWHEEL
```
No check-ins of commented sources

awkay commented on src/main/fulcrologic/semantic_ui/factories.cljs:
```diff
@@ -1,117 +1,92 @@
-    goog.object
-    [fulcro.client.dom :as dom]))
-
-(defn factory-apply
```
This is a library and these were public. You can deprecate them and make a new ns for them, but don't delete them. Make them refer to the new versions, just like you did for all of the factories.

awkay commented on src/main/fulcrologic/semantic_ui/modules/ui_accordion.cljs:
```diff
@@ -0,0 +1,63 @@
+(ns fulcrologic.semantic-ui.modules.ui-accordion
+  (:require
+    [fulcrologic.semantic-ui.factory-helpers :as h]
+    ["semantic-ui-react" :refer [Accordion AccordionPanel AccordionContent AccordionTitle AccordionAccordion]]))
```
Did you check that this reduces the advanced compile size?  You're doing a require against the top-level index.js, which will pull in everything.

I could be wrong, but I think you've got to drop to the module level on the require to get the size reduction.

```
npx shadow-cljs run shadow.cljs.build-report build-name report.html
```

I think the proper requires use the path to the module...something like this:

```
(ns fulcrologic.semantic-ui.modules.ui-button
  (:require
    [fulcrologic.semantic-ui.factory-helpers :as h]
    ["semantic-ui-react/dist/commonjs/elements/Button/Button" :default Button]
    ["semantic-ui-react/dist/commonjs/elements/Button/ButtonOr" :default ButtonOr]
    ["semantic-ui-react/dist/commonjs/elements/Button/ButtonGroup" :default ButtonGroup]
    ["semantic-ui-react/dist/commonjs/elements/Button/ButtonContent" :default ButtonContent]))
```

Use the report to see if that helps.  It is a bit more complexity in the code generation, but the whole point was to make if obvious what could be removed from the build.  To get the ideal smallest size I'd guess there would need to be a namespace per element/module.

I could be wrong on this.  My experience, though, is that the current implementation causes *all* of them to end up in an advanced build, and I think your changes will not change that.

NOTE: This change makes the library *only* work with shadow-cljs, since there is no such require statement in standard cljs.

awkay replied:
I did check this, and it does work as I describe.  Your changes will not affect build size unless you do the above.

I got 500K on your build when just using button, and 100k with my version (again just using button).  The other bad news is that even when I did it that way it pulled in a lot of cruft I didn't directly use...I'm guessing because some of those components use things like dimmer.

awkay commented:
Missing package.json for workspaces build.  Did you forget to check it in?

j1mr10rd4n commented:
Superseded by work on [optimise-for-shadow-cljs branch](https://github.com/fulcrologic/semantic-ui-wrapper/tree/optimise-for-shadow-cljs).

danieroux committed:
Make compatible with ClojureScript (not only shadow-cljs)

- Change factory-ns generation to use library property namespaces
- Above makes this lib usable with ClojureScript from 1.10.844 onwards: https://clojurescript.org/news/2021-04-06-release#_library_property_namespaces
- Use the lowest shadow-cljs that still loads from central mvn repo (2.12.6)
